File: day12/EmpMan_phase2/tests/pages/LoginPage.js
/**
 * Page Object Model – Login Page
 * Covers: Admin tab, Employee tab, username/password inputs,
 *         toggle password visibility, error messages, submit.
 */
class LoginPage {
  constructor(page) {
    this.page = page;

    // --- Tabs ---
    this.adminTab    = page.locator('#admin-login-tab');
    this.employeeTab = page.locator('#employee-login-tab');

    // --- Form fields ---
    this.usernameInput = page.locator('#login-username');
    this.passwordInput = page.locator('#login-password');
    this.submitBtn     = page.locator('#login-submit-btn');
    this.errorMsg      = page.locator('.login-error');

    // --- Brand / UI elements ---
    this.brandTitle    = page.locator('.login-brand-title');
    this.loginTitle    = page.locator('.login-title');
    this.demoCredsBox  = page.locator('.login-demo-creds');
  }

  /** Navigate to the app root */
  async goto() {
    await this.page.goto('/');
    await this.page.waitForLoadState('networkidle');
  }

  /** Select the Admin tab */
  async selectAdminTab() {
    await this.adminTab.click();
  }

  /** Select the Employee tab */
  async selectEmployeeTab() {
    await this.employeeTab.click();
  }

  /**
   * Perform a full login
   * @param {'admin'|'employee'} role
   * @param {string} username
   * @param {string} password
   */
  async login(role, username, password) {
    if (role === 'admin') {
      await this.selectAdminTab();
    } else {
      await this.selectEmployeeTab();
    }
    await this.usernameInput.fill(username);
    await this.passwordInput.fill(password);
    await this.submitBtn.click();
  }

  /** Wait until the Dashboard is visible (post-login redirect) */
  async waitForDashboard() {
    await this.page.waitForSelector('.dashboard', { timeout: 15000 });
  }

  /** Wait until the ESS Dashboard is visible (employee login) */
  async waitForESSDashboard() {
    await this.page.waitForSelector('.ess-dashboard, [id="nav-ess-dashboard"]', { timeout: 15000 });
  }

  /** Retrieve visible error message text */
  async getErrorText() {
    await this.errorMsg.waitFor({ state: 'visible', timeout: 8000 });
    return this.errorMsg.innerText();
  }

  /** Check if submit button is in loading state */
  async isLoading() {
    return (await this.submitBtn.innerText()).includes('Signing');
  }
}

module.exports = { LoginPage };

File: day12/EmpMan_phase2/tests-java/src/main/java/com/ems/pages/LoginPage.java
package com.ems.pages;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import com.microsoft.playwright.options.AriaRole;

public class LoginPage {
    private final Page page;
    
    public final Locator container;
    public final Locator adminTab;
    public final Locator employeeTab;
    public final Locator usernameInput;
    public final Locator passwordInput;
    public final Locator loginBtn;
    public final Locator errorMsg;
    public final Locator dashboardTitle;

    public LoginPage(Page page) {
        this.page = page;
        this.container = page.locator(".login-container, .login-page");
        this.adminTab = page.locator("button").filter(new Locator.FilterOptions().setHasText("Admin Login"));
        this.employeeTab = page.locator("button").filter(new Locator.FilterOptions().setHasText("Employee Login"));
        this.usernameInput = page.locator("input[type='text'], input[name='username']");
        this.passwordInput = page.locator("input[type='password'], input[name='password']");
        this.loginBtn = page.locator("button").filter(new Locator.FilterOptions().setHasText("Sign In"));
        this.errorMsg = page.locator(".error-message");
        this.dashboardTitle = page.locator(".dashboard, .sidebar");
    }

    public void navigateToRoot() {
        page.navigate("/");
        page.waitForLoadState(com.microsoft.playwright.options.LoadState.NETWORKIDLE);
    }

    public void login(String username, String password) {
        usernameInput.fill(username);
        passwordInput.fill(password);
        loginBtn.click();
    }

    public void selectAdminTab() {
        if(adminTab.isVisible()) adminTab.click();
    }

    public void selectEmployeeTab() {
        if(employeeTab.isVisible()) employeeTab.click();
    }
}
